src/bun.js/api/BunObject.zig

```diff
@@ -2563,16 +2597,154 @@ pub const Crypto = struct {
 
             const output_digest_slice: []u8 = &output_digest_buf;
 
+            switch (this.*) {
+                .evp => {},
+                .zig => |*inner| {
+                    inner.final(output_digest_slice);
+                    const digest_length = inner.digest_length();
+                    return encoding.encodeWithMaxSize(globalThis, BoringSSL.EVP_MAX_MD_SIZE, output_digest_slice[0..digest_length]);
+                },
+            }
+
             const out = this.evp.final(globalThis.bunVM().rareData().boringEngine(), output_digest_slice);
 
             return encoding.encodeWithMaxSize(globalThis, BoringSSL.EVP_MAX_MD_SIZE, out);
         }
 
         pub fn finalize(this: *CryptoHasher) callconv(.C) void {
-            // https://github.com/oven-sh/bun/issues/3250
-            this.evp.deinit();
+            switch (this.*) {
+                .evp => {
+                    // https://github.com/oven-sh/bun/issues/3250
+                    this.evp.deinit();
+                },
+                .zig => |*inner| {
+                    inner.deinit();
+                },
+            }
+            this.destroy();
+        }
+    };
 
-            bun.default_allocator.destroy(this);
+    const CryptoHasherZig = struct {
+        algorithm: EVP.Algorithm,
+        state: *anyopaque,
```

Comment:
Using `*anyopaque` here is more unsafe than necessary.

if the BoringSSL structs are large enough, we could instead use a `union(enum)` here of each supported zig crypto hasher type. From there, we can avoid the extra allocation for the state and the lifetime of this is simpler.

If the BoringSSL types are not already large, then we could use a `bun.TaggedPointerUnion`  and that will store the expected type in the same size as a regular pointer

Reply:
```zig
pub const CryptoHasher = union(enum) {
    evp: EVP,
    zig: CryptoHasherZig,
```

`evp` is `Algorithm + ptr + (ptr + ptr + ptr + ptr)`
`zig` is `Algorithm + ptr` as-is

`@sizeOf(std.crypto.hash.sha3.Sha3_*)` is `352`, `344`, `312`, `280`.

Reply:
ah thats kind of large

can we do TaggedPointerUnion?